app/src/main/java/com/example/farmeraid/data/MarketRepository.kt

```diff
@@ -2,69 +2,99 @@ package com.example.farmeraid.data
+        }
+
+        val marketModel : MutableList<DocumentSnapshot> = mutableListOf()
+        marketIds.data?.forEach {
+            try {
+                marketModel.add(db.collection("market").document(it).get().await())
+            } catch (e : Exception) {
+                return ResponseModel.FAResponseWithData.Error(e.message ?: "Unknown error while fetching market")
+            }
+        }
+
+        val marketModelList = mutableListOf<MarketModel.MarketWithQuota>()
+
+        for (market in marketModel) {
+            quotasRepository.getQuota(market.id)?.let {
+                MarketModel.MarketWithQuota(
+                    id = market.id,
+                    name = market.get("name") as String,
+                    quota = it
+                )
+            }?.let { marketModelList.add(it) }
+        }
+
+        return ResponseModel.FAResponseWithData.Success(marketModelList)
+    }
+
+    suspend fun getMarkets(): Flow<ResponseModel.FAResponseWithData<List<MarketModel.Market>>> {
         return flow {
-            emit(markets)
+            emit(readMarketData())
         }.flowOn(Dispatchers.IO)
     }
 
-    fun getMarketsWithQuota() : Flow<List<MarketModel.MarketWithQuota>> {
+    suspend fun getMarketsWithQuota() : Flow<ResponseModel.FAResponseWithData<List<MarketModel.MarketWithQuota>>> {
         return flow {
             emit(
-                markets.mapNotNull { market ->
-                    quotasRepository.getQuota(market.id)
-                        ?.let { quota ->
-                            MarketModel.MarketWithQuota(
-                                id = market.id,
-                                name = market.name,
-                                quota = quota,
-                            )
-                        }
-                }
+                readMarketDataWithQuotas()
             )
         }
     }
 
-    fun getMarket(id: String) : MarketModel.Market? {
-        return markets.firstOrNull { it.id == id }
-    }
-
-    suspend fun getMarketWithQuota(id : String) : MarketModel.MarketWithQuota? {
-        return markets.firstOrNull { it.id == id }
-            ?.let { market ->
-                quotasRepository.getQuota(market.id)
-                    ?.let { quota ->
-                        MarketModel.MarketWithQuota(
-                            id = market.id,
-                            name = market.name,
-                            quota = quota,
-                        )
-                    }
-            }
+    suspend fun getMarketWithQuota(id : String) : ResponseModel.FAResponseWithData<MarketModel.MarketWithQuota> {
+        return readMarketDataWithQuotas().let { marketWithQuotas ->
+            marketWithQuotas.data?.let { it ->
+                ResponseModel.FAResponseWithData.Success(it.firstOrNull { it.id == id })
+            } ?: ResponseModel.FAResponseWithData.Error(marketWithQuotas.error ?: "Unkown error reading quotas")
```

Comment:
Actually I just realized. For the getMarketWithQuota function, the point is that we shouldn't have to read all the market with quotas to get one. Its so that we can just fetch for the one that we want so it takes less time. So in this function, we should be using the readMarketDataWithQuotas() function. We should be doing a db fetch for the single market and quota associated with the id thats passed in.

Reply:
True, I'll create a function to view an individual quota

Reply:
You don't have to create a new one, just change the getMarketWithQuota implementation to get an individual quota. Since that was its indended purpose.

Reply:
This one calls the readMarketDataWithQuotas function, which gets all the markets. So would it be better to have another function to just call firestore to fetch one?